(CODEMGMT-1415) Sundry Improvements

justinstoller opened the pull request:
* Fixes regression where on some module content would be deleted on subsequent runs when not using `--modules`
* Renames `--assume-unchanged` flag to `--incremental`
* Adds CHANGELOG entry

Still to do
* [ ] Acceptance test

justinstoller committed:
(CODEMGMT-1415) Ensure environment purge_exclusions are calculated

In commit 9d0579f07 Environments stopped interrogating their Puppetfile
for managed content needed to calculate their purge exclusions, but
moved to using the data returned by the ModuleLoader when loading
content.

However, if content was not loaded prior to attempting to purge the
environment then the environment would purge much of the Puppetfile
managed content. This would happen when running `r10k deploy
environment` (without `--modules`) on an existing environment.

To resolve this we now conditionally load module data if it hasn't been
loaded yet when resolving our purge exclusions.

justinstoller committed:
(CODEMGMT-1415) Use `--incremental` instead of `--assume-unchanged`

Concerns were raised that `--assume-unchanged` required knowing too much
about how R10K internals work and that users would be well aware of this
kind of behavior from back up tools where the `--incremental` is the
standard terminology.

justinstoller committed:
(CODEMGMT-1415) Update CHANGELOG

Magisus reviewed (approved):
I think the first commit could use a short description of the bug.

Are you planning to add the acceptance test as part of this PR, or should we merge this first to unblock code-manager's unit tests?

mwaggett commented on spec/unit/action/deploy/environment_spec.rb:
```diff
@@ -147,11 +147,11 @@
       end
 
-      it "assume unchanged flag causes the module definitons to be preloaded by the loader" do
+      it "incremental unchanged flag causes the module definitons to be preloaded by the loader" do
```
```suggestion
      it "incremental flag causes the module definitons to be preloaded by the loader" do
```

justinstoller replied:
smh, I guess it really shows that my search term was just "assume"

mwaggett reviewed (commented):
+1 to updating the commit message to describe the bug

justinstoller committed:
Update spec/unit/action/deploy/environment_spec.rb

Co-authored-by: Molly Waggett <mwaggett@alumni.reed.edu>

justinstoller committed:
(CODEMGMT-1415) Basic acceptance test for environment deploys

When evaluating how to test `--incremental` behavior it was decided
that the existing acceptance tests lacked basic sanity tests (there was
an uncaught bug when running a simple `r10k deploy environment` twice).

This provides basic acceptance level testing for running `r10k deploy
environment` with and without `--modules` and `--incremental`.

justinstoller committed:
(CODEMGMT-1415) Move Puppetfile logging

When implementing and refactoring the Puppetfile ModuleLoader logging
that happened on instantiation of the Puppetfile was moved to the
instantiation of the ModuleLoader. However, the a ModuleLoader is
instantiated for both the legacy Puppetfile as well as the owning
Environment directly, causing the log lines to be emitted twice.

Magisus reviewed (approved):
I think this looks good so long as the test passes. It checks all the boxes I would want to see for the basic workflow.

mwaggett commented on integration/tests/basic_functionality/basic_deployment.rb:
```diff
@@ -0,0 +1,144 @@
+end
+
+step 'Test --modules --incremental updates changed modules' do
+  on(master, "#{r10k_fqp} deploy environment production --modules --incremental --verbose=debug1") do |result|
```
Without a new Puppetfile, this deploy isn't actually changing anything, is it? Is there a better way we can test this behavior other than checking the log messages? Just trying to think in terms of actual workflows, since I don't imagine someone would run a deploy with `--modules` and then again with `--modules --incremental` without at least making some sort of change in between?

justinstoller replied:
Yeah, I struggled with this. The end result of running `--modules` vs `--modules --incremental` should be the same, modules that should be updated will be updated. However, the how it happens is different, with some modules being skipped completely and some still being checked. So the logging shows us at least that the static version checking is correct (for a forge module) and that unchanged static modules are skipped.

If you think I should I could include all the permutations of different types of versions as well as those that have changed definitions? Or are you mainly concerned about those modules whose definition changed on disk part? I could include one of those in the Puppetfile.

justinstoller replied:
I'll add a module that is updated in this invocation to test the changeable vs static, changed vs static, unchanged. Let me know if you'd like more permutations of static version checking (fwiw, there is some unit tests around that: https://github.com/puppetlabs/r10k/commit/6f46384633531d82df10d490183259eaccd8ad93#diff-1bc365ddca4e6e4a7ba95ff4d655c036faa16ec1025e35bc559476265a5719aaR344-R389)

mwaggett replied:
I think what you added looks good! I'm satisfied with the various permutations we test here.

mwaggett commented on integration/tests/basic_functionality/basic_deployment.rb:
```diff
@@ -0,0 +1,146 @@
+
+step 'Test second run of deploy updates control repo, but not modules' do
+  puppetfile2 =<<-EOS
+    mod 'puppetlabs/apache', :latest
```
Might be good to add a comment on what the latest version is at the time of writing this. Could help with future debugging and also explains why `apache_major_version > 5` is an appropriate check.

justinstoller committed:
(CODEMGMT-1415) Improve acceptance test

mwaggett reviewed (approved):
![spongebob giving many thumbs up](https://media.giphy.com/media/3o7abGQa0aRJUurpII/giphy.gif)
